src/Coordination/Changelog.cpp

```diff
@@ -1014,8 +1016,57 @@ void Changelog::writeThread()
 {
     WriteOperation write_operation;
     bool batch_append_ok = true;
-    while (write_operations.pop(write_operation))
+    size_t pending_appends = 0;
+    bool try_batch_flush = false;
+
+    const auto flush_logs = [&](const auto & flush)
+    {
+        LOG_TEST(log, "Flushing {} logs", pending_appends);
+
+        {
+            std::lock_guard writer_lock(writer_mutex);
+            current_writer->flush();
+        }
+
+        {
+            std::lock_guard lock{durable_idx_mutex};
+            last_durable_idx = flush.index;
+        }
+
+        pending_appends = 0;
+    };
+
+    const auto notify_append_completion = [&]
     {
+        durable_idx_cv.notify_all();
+
+        // we need to call completion callback in another thread because it takes a global lock for the NuRaft server
+        // NuRaft will in some places wait for flush to be done while having the same global lock leading to deadlock
+        // -> future write operations are blocked by flush that cannot be completed because it cannot take NuRaft lock
+        // -> NuRaft won't leave lock until its flush is done
+        if (!append_completion_queue.push(batch_append_ok))
+            LOG_WARNING(log, "Changelog is shut down");
+    };
+
+    while (true)
+    {
+        if (try_batch_flush)
+        {
+            try_batch_flush = false;
+            if (!write_operations.tryPop(write_operation))
```

Comment:
It's really complex! So the logic is as follows:
When we encounter flush operation we don't execute it immediately, but set flag `try_batch_flush = true` and skip operation, but store in `write_operation` variable. On the next iteration we will try to process next `write_operation` but if we don't have anything -- we will execute `flush` this pending.

Reply:
BTW here it may make sense to `tryPop` with some timeout, which is smaller than actual flush time. It should depend on block device, but maybe add configuration setting with 0 by default.

Reply:
I would avoid timeout simply because the cost of condition variable wait can be much higher than any potential gain of batching flushes.
I can try to introduce later on some setting which will also try to wait for some time instead of just limiting by number of logs we want to flush.